### PR TITLE
keep min package versions from deletion

### PR DESCRIPTION
**Pull request by NamrataJha**
This PR adds 
- A `min-versions-to-keep` option for setting a lower limit on package versions while deleting. 
- A `ignore-versions` option to set regex for packages to be ignored while deletion.
- A `delete-only-pre-release-versions` option to delete only the pre-release versions for a package
- This integrates the changes from [PR #29](https://github.com/actions/delete-package-versions/pull/29), [PR #10](https://github.com/actions/delete-package-versions/pull/10)  and resolves additional conflicts on those branches.

## New Feature
- Specify a minimum number of latest package versions to always keep while deleting.
  - Use the `min-versions-to-keep` attribute with the action workflow.
  - This is not a mandatory field for running the action. When not specified, by default the value is 0 and at every run `num-old-versions-to-delete` package versions will get deleted.
  - When `min-versions-to-keep` is set to any value greater than 0, it will delete maximum possible package versions except `min-versions-to-keep` package versions.
- Specify package versions to ignore while deletion by regex pattern.
  - Use the `ignore-versions` attribute within the action workflow.
  - This is not a mandatory field. By default, no package version will be ignored.
- Set to delete only pre-release versions.
  - Use the `delete-only-pre-release-versions` attribute with the action workflow.
  - This is not a mandatory field for running the action. By default, this is set to `"false"`.
  - When `"true"`, number of pre-release versions to keep can be set using `min-versions-to-keep` with this option.

## What to Test
- Test using the latest commit of delete-package-versions action - `actions/delete-package-versions@6fee3def5eefbcb45b0ab364630c00099189e727` to use the changes.
- Specify value for `min-versions-to-keep` `ignore-versions` and `delete-only-pre-release-versions` and see if the behaviour is expected.

### CODE REVIEW

**Comment by nadesu** [comment]
@NamrataJha - pls add more details to description as to how the new features of Action can be used, its spec, etc.

**Comment by nadesu** [comment]
@NamrataJha - pls work with @jcansdale for any feedback

File: README.md [hunk 1]
```diff
@@ -197,6 +204,48 @@ Delete the oldest 3 version of a package hosted in a different repo than the one
...
+__Example__
+
+Delete the oldest 3 version and always keep atleast 2 versions of a package hosted in the same repo as the workflow
+
+```yaml
+- uses: actions/delete-package-versions@v1
+  with:
+    package-name: 'test-package'
+    num-old-versions-to-delete: 3
+    min-versions-to-delete: 2
```
Comment (nishthaGupta) [inline README.md:222]: @NamrataJha Shouldn't it be _min-versions-to-keep_ here?

Comment (NamrataJha) [reply README.md:222]: Yeah sorry my bad. Will fix the README.md

**Review by nishthaGupta — APPROVED** [review]
Approving.

**Comment by NamrataJha** [comment]
Have added an additional option `ignore-versions` to filter out packages to be deleted.

File: action.yml [hunk 2]
```diff
@@ -34,6 +34,19 @@ inputs:
...
+      Number of versions to keep starting with the latest version
+    required: false
+    default: "1"
+
+  ignore-versions:
+    description: >
+      Comma seperated regex patterns of versions to ignore.
+      Defaults to all major and minor versions and only deletes pre-release and patch versions.
+    required: false
+    default: "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0+)$"
```
Comment (jcansdale) [inline action.yml:48]: NuGet is pretty loose about what version strings are valid. See here for more info:
[I think NuGet users ](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#where-nugetversion-diverges-from-semantic-versioning)

In particular:

> `NuGetVersion` only requires the major segment to be defined. All others are optional, and are equivalent to zero. This means that `1`, `1.0`, `1.0.0`, and `1.0.0.0` are all accepted and equal.

I'd be include to default to keeping all versions that only include number segments.

That way we wouldn't delete `1`, `1.0`, `1.0.0`, and `1.0.0.0`, but we would delete `1.0.0-foo` and `1.0.0.pre.foo` (Ruby style).

Does that make sense?

**Review by jcansdale — CHANGES_REQUESTED** [review]
I think we should be a little more conservative about what versions we default to keeping. See inline comment for suggestion.

File: action.yml (hunk 2 above)
Comment (NamrataJha) [reply action.yml]: This does makes sense as we are looking to delete only pre-release versions. Updating the regex to include versions like `1` `1.0` `1.0.0` `1.0.0.0` and so on. Ignoring rest of the package versions like `1.0.0-foo` and others.

File: src/delete.ts [hunk 3]
```diff
@@ -13,9 +13,20 @@ export function getVersionIds(input: Input): Observable<string[]> {
...
       input.token
-    ).pipe(map(versionInfo => versionInfo.map(info => info.id)))
+    ).pipe(
+      map(versionInfo => {
+        const numberVersionsToDelete =
+          versionInfo.length - input.minVersionsToKeep
+        return numberVersionsToDelete <= 0
+          ? []
+          : versionInfo
+              .filter(info => !input.ignoreVersions.test(info.version))
```
Comment (AmrutaKawade) [inline src/delete.ts:25]: What does this test method do?

File: README.md [hunk 4]
```diff
@@ -197,6 +214,48 @@ Delete the oldest 3 version of a package hosted in a different repo than the one
...
+The [token][token] needs the delete packages and read packages scope. It is recommended [to store the token as a secret][secret]. In this example the [token][token] was stored as a secret named __GITHUB_PAT__.
+
+__Example__
+
+Delete the oldest 3 version and always keep atleast 2 versions of a package hosted in a different repo than the workflow
+
+```yaml
+- uses: actions/delete-package-versions@v1
+  with:
+    owner: 'github'
```
Comment (AmrutaKawade) [inline README.md:250]: can you add example for ignore versions as well?

File: src/delete.ts (hunk 3 above)
Comment (NamrataJha) [reply src/delete.ts:25]: The test method checks for package versions against the regex provided for package versions to be ignored. If they match we filter them out.

File: README.md (hunk 4 above)
Comment (NamrataJha) [reply README.md:250]: Sure will add this.

**Comment by NamrataJha** [comment]
@jcansdale as per our discussion have updated to the action to-
- remove `ignore-versions` default value. No package version will be ignored by default.
- By default `min-versions-to-keep` is set to 1.
- By default `num-old-versions-to-delete` is set to 1.
- `delete-only-pre-release-versions` is the option for deleting all pre-release versions. This is `false` by default.
- When `delete-only-pre-release-versions` is set to `true`, `min-versions-to-keep` is set to 0, `num-old-versions-to-delete` is set to 100 and `ignore-versions` is set to `^(0|[1-9]\\d*)((\\.(0|[1-9]\\d*))*)$` which includes all the release versions.

@AmrutaKawade Have made the above changes in the action and updated the README.md. Please review this once.

File: src/input.ts [hunk 5]
```diff
@@ -32,7 +41,16 @@ export class Input {
     this.repo = validatedParams.repo
     this.packageName = validatedParams.packageName
     this.numOldVersionsToDelete = validatedParams.numOldVersionsToDelete
+    this.minVersionsToKeep = validatedParams.minVersionsToKeep
+    this.ignoreVersions = validatedParams.ignoreVersions
+    this.deletePreReleaseVersions = validatedParams.deletePreReleaseVersions
     this.token = validatedParams.token
+
+    if (this.deletePreReleaseVersions == 'true') {
+      this.numOldVersionsToDelete = 100
```
Comment (AmrutaKawade) [inline src/input.ts:50]: why we are setting it to 100? what if user wants to delete more versions. we are ignoring this input when deletePreReleaseVersions is set.

File: README.md [hunk 6]
```diff
@@ -197,6 +218,82 @@ Delete the oldest 3 version of a package hosted in a different repo than the one
...
+    token: ${{ secrets.GITHUB_PAT }}
+```
+
+<br>
+
+### Delete oldest x number of versions while ignoring particular package versions in the same repo as the workflow
+
+To delete oldest x number of versions while ignoring all the major package versions in the same repo as the workflow the __package-name__, __num-oldest-versions-to-delete__ and __ignore-versions__ inputs are required.
+
+There is a possibility if the oldest x number of versions contain ignored package versions, actual package versions to ge deleted will be less than x.
```
Comment (AmrutaKawade) [inline README.md:268]: typo in statement

File: src/input.ts (hunk 5 above)
Comment (NamrataJha) [reply src/input.ts:50]: The graphql api has a limit to get only 100 records. On increasing this limit the api call results in an error.

File: README.md (hunk 6 above)
Comment (NamrataJha) [reply README.md]: Fixed it. Thanks.

File: src/input.ts (hunk 5 above)
Comment (AmrutaKawade) [reply src/input.ts]: ok got it

File: src/input.ts [hunk 7]
```diff
@@ -32,7 +41,19 @@ export class Input {
     this.repo = validatedParams.repo
     this.packageName = validatedParams.packageName
     this.numOldVersionsToDelete = validatedParams.numOldVersionsToDelete
+    this.minVersionsToKeep = validatedParams.minVersionsToKeep
+    this.ignoreVersions = validatedParams.ignoreVersions
+    this.deletePreReleaseVersions = validatedParams.deletePreReleaseVersions
     this.token = validatedParams.token
+
+    if (this.minVersionsToKeep > 0) {
+      this.numOldVersionsToDelete = 100 - this.minVersionsToKeep
```
Comment (AmrutaKawade) [inline src/input.ts:50]: do we set limit to 100 if just `num-old-versions-to-delete` is provided and `min-versions-to-keep:` and `  delete-only-pre-release-versions:` are not provided?

File: src/version/get-versions.ts [hunk 8]
```diff
@@ -89,9 +89,10 @@ export function getOldestVersions(
   ).pipe(
     map(result => {
       if (result.repository.packages.edges.length < 1) {
-        throwError(
+        console.log(
```
Comment (AmrutaKawade) [inline src/version/get-versions.ts:92]: is this added for testing or are we going to swallow error?

File: src/input.ts (hunk 7 above)
Comment (NamrataJha) [reply src/input.ts:50]: The limit of 100 is for graphql api so we are adjusting the package versions to be queried around this. So we always maintain `num-old-versions-to-delete`+`min-versions-to-keep` = 100.

File: src/version/get-versions.ts (hunk 8 above)
Comment (NamrataJha) [reply src/version/get-versions.ts:92]: The throwError was not working as expected and breaking the action run. Hence we are giving the warning message and exiting the action without causing it to fail.

**Review by AmrutaKawade — APPROVED** [review]
LGTM

File: README.md [hunk 9]
```diff
@@ -37,8 +40,26 @@ This action deletes versions of a package from [GitHub Packages](https://github.
...
   num-old-versions-to-delete:
 
+  # The number of latest versions to not delete.
+  # Defaults to 0.
+  # When this is set greater than 0 it will delete all deletable package versions except the specified no.
+  # `num-old-versions-to-delete` will not be taken into account.
+  # Cannot be more than 100
+  min-versions-to-keep:
+
+  # The package versions to ignore exclude from deletion.
```
Comment (pulkitaggarwl) [inline README.md:53]: can we update the text to suggest that we are supporting regex here.

Comment (NamrataJha) [reply README.md:53]: Sure will make the change.

File: README.md [hunk 10]
```diff
@@ -49,13 +70,20 @@ This action deletes versions of a package from [GitHub Packages](https://github.
...
+  - [Delete all except y latest versions of a package hosted in the same repo as the workflow](#delete-all-except-y-latest-versions-of-a-package-hosted-in-the-same-repo-as-the-workflow)
+  - [Delete all except y latest versions of a package hosted in a different repo than the workflow](#delete-all-except-y-latest-versions-of-a-package-hosted-in-a-different-repo-than-the-workflow)
+  - [Delete oldest x number of versions while ignoring particular package versions in the same repo as the workflow](#delete-oldest-x-number-of-versions-while-ignoring-particular-package-versions-in-the-same-repo-as-the-workflow)
+  - [Delete all except y latest versions while ignoring particular package versions in the same repo as the workflow](#delete-all-except-y-latest-versions-while-ignoring-particular-package-versions-in-the-same-repo-as-the-workflow)
+  - [Delete only pre-release package versions except y no of pre-release package versions in the same repo as the workflow](#delete-only-pre-release-package-versions-except-y-no-of-pre-release-package-versions-in-the-same-repo-as-the-workflow)
```
Comment (jcansdale) [inline README.md:85]: Can we move these two to the top?

I suspect the current top one, `Delete a specific version of a package`, is rarely if ever used! 😉

**Review by jcansdale — CHANGES_REQUESTED** [review]
Suggestions to make the new functionality more prominent/discoverable.

**Pull request merged**